frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -0,0 +1,839 @@
+			desired_targets,
+			size,
+			T::MinerMaxWeight::get(),
+		);
+		log!(
+			debug,
+			"miner: current compact solution voters = {}, maximum_allowed = {}",
+			compact.voter_count(),
+			maximum_allowed_voters,
+		);
+		let compact = Self::trim_compact(maximum_allowed_voters, compact, &voter_index)?;
+
+		// re-calc score.
+		let winners = sp_npos_elections::to_without_backing(winners);
+		let score = compact.clone().score(&winners, stake_of, voter_at, target_at)?;
+
+		let round = Self::round();
+		Ok((RawSolution { compact, score, round }, size))
+	}
+
+	/// Get a random number of iterations to run the balancing in the OCW.
+	///
+	/// Uses the offchain seed to generate a random number, maxed with
+	/// [`Config::MinerMaxIterations`].
+	pub fn get_balancing_iters() -> usize {
+		match T::MinerMaxIterations::get() {
+			0 => 0,
+			max @ _ => {
+				let seed = sp_io::offchain::random_seed();
+				let random = <u32>::decode(&mut TrailingZeroInput::new(seed.as_ref()))
+					.expect("input is padded with zeroes; qed")
+					% max.saturating_add(1);
+				random as usize
+			}
+		}
+	}
+
+	/// Greedily reduce the size of the a solution to fit into the block, w.r.t. weight.
+	///
+	/// The weight of the solution is foremost a function of the number of voters (i.e.
+	/// `compact.len()`). Aside from this, the other components of the weight are invariant. The
+	/// number of winners shall not be changed (otherwise the solution is invalid) and the
+	/// `ElectionSize` is merely a representation of the total number of stakers.
+	///
+	/// Thus, we reside to stripping away some voters. This means only changing the `compact`
+	/// struct.
+	///
+	/// Note that the solution is already computed, and the winners are elected based on the merit
+	/// of the entire stake in the system. Nonetheless, some of the voters will be removed further
+	/// down the line.
+	///
+	/// Indeed, the score must be computed **after** this step. If this step reduces the score too
+	/// much, then the solution will be discarded.
+	pub fn trim_compact<FN>(
+		maximum_allowed_voters: u32,
+		mut compact: CompactOf<T>,
+		nominator_index: FN,
+	) -> Result<CompactOf<T>, MinerError>
+	where
+		for<'r> FN: Fn(&'r T::AccountId) -> Option<CompactVoterIndexOf<T>>,
+	{
+		match compact.voter_count().checked_sub(maximum_allowed_voters as usize) {
+			Some(to_remove) if to_remove > 0 => {
+				// grab all voters and sort them by least stake.
+				let RoundSnapshot { voters, .. } =
+					Self::snapshot().ok_or(MinerError::SnapshotUnAvailable)?;
+				let mut voters_sorted = voters
+					.into_iter()
+					.map(|(who, stake, _)| (who.clone(), stake))
+					.collect::<Vec<_>>();
+				voters_sorted.sort_by_key(|(_, y)| *y);
+
+				// start removing from the least stake. Iterate until we know enough have been
+				// removed.
+				let mut removed = 0;
+				for (maybe_index, _stake) in
+					voters_sorted.iter().map(|(who, stake)| (nominator_index(&who), stake))
```

Comment:
I think the naming is missleading here and above, nominator_index must be voter_index, it can be a nominator or a validator.

Also this makes me think trim_compact can remove a validator thus making the winner set in `prepare_election_result` potentially wrong.

Reply:
I tried to avoid using validator and nominator in this pallet. We only know voters and targets in this stage. Will fix.